components/security-mgt/org.wso2.carbon.security.mgt/src/main/java/org/wso2/carbon/security/keystore/service/IdentityKeyStoreGeneratorImpl.java

```diff
@@ -0,0 +1,334 @@
+ */
+
+package org.wso2.carbon.security.keystore.service;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.bouncycastle.asn1.x500.X500Name;
+import org.bouncycastle.asn1.x509.SubjectPublicKeyInfo;
+import org.bouncycastle.cert.X509v3CertificateBuilder;
+import org.bouncycastle.cert.jcajce.JcaX509CertificateConverter;
+import org.bouncycastle.operator.jcajce.JcaContentSignerBuilder;
+import org.wso2.carbon.base.ServerConfiguration;
+import org.wso2.carbon.core.util.CryptoUtil;
+import org.wso2.carbon.core.util.KeyStoreManager;
+import org.wso2.carbon.identity.core.util.IdentityTenantUtil;
+import org.wso2.carbon.security.keystore.KeyStoreManagementException;
+import org.wso2.carbon.utils.ServerConstants;
+import org.wso2.carbon.utils.security.KeystoreUtils;
+
+import java.io.ByteArrayOutputStream;
+import java.math.BigInteger;
+import java.security.KeyPair;
+import java.security.KeyPairGenerator;
+import java.security.KeyStore;
+import java.security.PrivateKey;
+import java.security.SecureRandom;
+import java.security.cert.X509Certificate;
+import java.util.Date;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantLock;
+
+import static org.wso2.carbon.security.SecurityConstants.KeyStoreMgtConstants.KEY_STORE_CONTEXT_SEPARATOR;
+
+/**
+ * Implementation of the {@link IdentityKeyStoreGenerator} interface for generating and managing
+ * context-specific tenant key stores. This class provides functionality to create, manage, and store
+ * keystores in a multi-tenant environment, adhering to specific cryptographic requirements.
+ *
+ * <p><b>Key Features:</b></p>
+ * <ul>
+ *   <li>Generates keystores for tenants dynamically.</li>
+ *   <li>Supports various cryptographic algorithms and key generation techniques.</li>
+ *   <li>Handles secure persistence of keystores using {@link KeyStoreManager}.</li>
+ *   <li>Ensures thread-safe operations with locking mechanisms for multi-tenant environments.</li>
+ *   <li>Provides explainable methods for certificate creation, storage, and retrieval.</li>
+ * </ul>
+ *
+ * <p><b>Concurrency:</b></p>
+ * This implementation uses {@link ReentrantLock} to handle concurrent access to keystores for
+ * specific tenants, ensuring thread safety and data integrity.
+ *
+ * <p><b>Usage:</b></p>
+ * This class is intended to be used in environments where context-specific cryptographic needs
+ * must be met dynamically.
+ *
+ * <p><b>Exceptions:</b></p>
+ * The methods in this class throw {@link KeyStoreManagementException} for errors encountered during
+ * keystore creation, management, or persistence.
+ */
+public class IdentityKeyStoreGeneratorImpl implements IdentityKeyStoreGenerator{
+
+    private static final Log LOG = LogFactory.getLog(IdentityKeyStoreGeneratorImpl.class);
+    private String tenantDomain;
+    private String password;
+    private KeyStoreManager keyStoreManager;
```

Comment:
This class is used as a singleton through OSGI therefore lets not use class level variables. These seemed be initialized through the generateContextKeyStore method which is specific to each tenant. Lets remove the class level variables and pass these as param to required methods

Reply:
addressed in https://github.com/wso2/carbon-identity-framework/pull/6189/commits/6ca0b3d01b63c8f3f8ee122988af923b204f342b